working/2364 - primary constructors/feature-specification.md

```diff
@@ -609,17 +680,59 @@ Object {}`, and all three of them have `Object` as their direct superclass.*
-desugared using the same steps. This determines the dynamic semantics of a
-primary constructor.
-
-The following errors apply to formal parameters of a primary constructor.
-Let _p_ be a formal parameter of a primary constructor in a class `C`:
+Consider a class, enum, or extension type declaration _D_ with a primary
+header constructor *(note that it cannot be a `<mixinApplicationClass>`,
+because that kind of declaration does not support primary constructors,
+that is a syntax error)*. This declaration is treated as a class, enum,
+respectively extension type declaration without a primary header
+constructor which is obtained as described in the following. This
+determines the dynamic semantics of a primary constructor.
+
+A compile-time error occurs if the body of _D_ contains a non-redirecting
+generative constructor. *This ensures that every constructor invocation
+for this class will invoke the primary header constructor, either directly
+or via a series of generative redirecting constructors. This is required in
+order to allow initializers with no access to `this` to use the
+parameters.*
+
+The declaring parameter list of the primary header constructor introduces a
+new scope, the _declaring parameter scope_, whose enclosing scope is the
+type parameter scope of _D_, if any, and otherwise the enclosing library
+scope. The body scope of _D_ has the declaring parameter scope as its
+enclosing scope.
+
+*This implies that every parameter of the primary header constructor is in
+scope in the class body, unless the class body has a declaration with the
+same name (which would shadow the parameter).*
+
+A compile-time error occurs if an identifier resolves to a primary header
+constructor parameter, unless the identifier occurs in an initializing
+expression of a non-late instance variable declaration.
+
+*We can only use these parameters when it is guaranteed that the primary
+header constructor is currently being executed.*
+
+Assume that the primary header constructor has a declaring parameter with
+the name `n`, or an initializing formal parameter with the name `n`. Assume
+that an initializing expression of a non-late instance variable contains an
+identifier expression of the form `n` which is resolved as a reference to
+the instance variable which is initialized by said parameter *(that
+instance variable must also have the name `n`)*. In this case, the
+identifier expression evaluates to the value of the parameter.
+
+*This means that initializing expressions can, apparently, use the value of
+instance variables declared by the same class (not, e.g., inherited ones).
+They will actually get the value of the primary header parameter, but this
+value is also guaranteed to be the initial value of the corresponding
+instance variable.*
+
+*Note that it only applies to identifier expressions. In particular, this
+does not allow initializing expressions to assign to other instance
+variables.*
```

Comment:
It would be good to spell out how this behaves with respect to capture:

```dart
class C(var int x) {
  var closure = () => x;
}

main() {
  var c = C(1);
  c.x = 2;
  print(c.closure()); // Prints 1.
}
```

Reply:
Done; I put the example into the section where it is recommended that this situation causes a warning.